src/BSON/UTCDateTime.c

```diff
@@ -161,49 +157,58 @@ static bool php_phongo_utcdatetime_init_from_date(php_phongo_utcdatetime_t *inte
+   current time. */
 PHP_METHOD(UTCDateTime, __construct)
 {
 	php_phongo_utcdatetime_t    *intern;
 	zend_error_handling          error_handling;
-	zval                        *datetime = NULL;
+	zval                        *milliseconds = NULL;
 
 	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling TSRMLS_CC);
 	intern = Z_UTCDATETIME_OBJ_P(getThis());
 
-	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS() TSRMLS_CC, "|o!", &datetime) == SUCCESS) {
-		if (datetime == NULL) {
-			php_phongo_utcdatetime_init_from_current_time(intern);
-		} else if (instanceof_function(Z_OBJCE_P(datetime), php_date_get_date_ce() TSRMLS_CC)) {
-			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "|z!", &milliseconds) == FAILURE) {
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+	zend_restore_error_handling(&error_handling TSRMLS_CC);
+
+	if (milliseconds == NULL) {
+		php_phongo_utcdatetime_init_from_current_time(intern);
+		return;
+	}
+
+	if (Z_TYPE_P(milliseconds) == IS_OBJECT) {
+		if (instanceof_function(Z_OBJCE_P(milliseconds), php_date_get_date_ce() TSRMLS_CC)) {
+			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(milliseconds));
 #if PHP_VERSION_ID >= 50500
-		} else if (instanceof_function(Z_OBJCE_P(datetime), php_date_get_immutable_ce() TSRMLS_CC)) {
-			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+		} else if (instanceof_function(Z_OBJCE_P(milliseconds), php_date_get_immutable_ce() TSRMLS_CC)) {
+			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(milliseconds));
 #endif
 		} else {
-			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected instance of DateTimeInterface, %s given", ZSTR_VAL(Z_OBJCE_P(datetime)->name));
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected instance of DateTimeInterface, %s given", ZSTR_VAL(Z_OBJCE_P(milliseconds)->name));
 		}
-
-		zend_restore_error_handling(&error_handling TSRMLS_CC);
 		return;
 	}
 
-	{
-		char                *s_milliseconds;
-		phongo_zpp_char_len  s_milliseconds_len;
-
-		if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &s_milliseconds, &s_milliseconds_len) == FAILURE) {
-			zend_restore_error_handling(&error_handling TSRMLS_CC);
-			return;
-		}
+	if (Z_TYPE_P(milliseconds) == IS_DOUBLE) {
+		convert_to_long(milliseconds);
+	}
 
-		php_phongo_utcdatetime_init_from_string(intern, s_milliseconds, s_milliseconds_len TSRMLS_CC);
+	if (Z_TYPE_P(milliseconds) == IS_LONG) {
+		php_phongo_utcdatetime_init(intern, Z_LVAL_P(milliseconds));
+		return;
 	}
 
-	zend_restore_error_handling(&error_handling TSRMLS_CC);
+	if (Z_TYPE_P(milliseconds) != IS_STRING) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected integer or string, %s given", zend_get_type_by_const(Z_TYPE_P(milliseconds)));
```

Comment:
HHVM only throws "Expected instance of DateTimeInterface" and "Error parsing %s as 64-bit integer" exceptions, so it may need a PR for this (see [here](https://github.com/mongodb/mongo-hhvm-driver/blob/d36a92128a782ff2b678127ff745bcbfc78a05a9/ext_mongodb.php#L1721)).